core/src/main/java/org/apache/carbondata/core/util/CleanFilesUtil.java

```diff
@@ -0,0 +1,205 @@
+      SegmentFileStore fileStore = new SegmentFileStore(carbonTable.getTablePath(),
+          staleSegmentFile);
+      List<String> filesToProcess = fileStore.readIndexFiles(SegmentStatus.SUCCESS, true,
+          FileFactory.getConfiguration());
+
+      // get carbondata files from here
+      Map<String, List<String>> indexFilesMap = fileStore.getIndexFilesMap();
+      for (Map.Entry<String, List<String>> entry : indexFilesMap.entrySet()) {
+        filesToProcess.addAll(entry.getValue());
+      }
+      // After all the files have been added to list, move them to the trash folder
+      TrashUtil.copyFilesToTrash(filesToProcess, TrashUtil.getCompleteTrashFolderPath(
+          carbonTable.getTablePath(), timeStampForTrashFolder, segmentNumber), segmentNumber);
+      // After every file of that segment has been copied, need to delete those files.
+      try {
+        for (String file : filesToProcess) {
+          FileFactory.deleteFile(file);
+        }
+        // Delete the segment file too
+        FileFactory.deleteFile(CarbonTablePath.getSegmentFilePath(carbonTable.getTablePath(),
+            staleSegmentFile));
+        // remove duplicate segment files if any
+        for (String duplicateStaleSegmentFile : redundantSegmentFile) {
+          if (DataFileUtil.getSegmentNoFromSegmentFile(duplicateStaleSegmentFile)
+              .equals(segmentNumber)) {
+            FileFactory.deleteFile(CarbonTablePath.getSegmentFilePath(carbonTable.getTablePath(),
+                duplicateStaleSegmentFile));
+          }
+        }
+      } catch (IOException e) {
+        LOGGER.error("Error while deleting the source data files. Please delete the files of" +
+            " segment: " + segmentNumber + " manually.", e);
+      }
+    }
+    // Delete the empty partition folders
+    deleteEmptyPartitionFoldersRecursively(FileFactory.getCarbonFile(carbonTable.getTablePath()));
+  }
+
+  /**
+   * This method will find all the stale segments by comparing the segment files in the
+   * metadata directory with the segments in the table status file. Any segment which has entry
+   * in the metadata folder and is not present in the table status file is considered as a
+   * stale segment. Only comparing from tablestatus file, not checking tablestatus.history file
+   */
+  private static void getStaleSegmentFiles(CarbonTable carbonTable, List<String> staleSegmentFiles,
+      List<String> redundantSegmentFile) {
+    String segmentFilesLocation =
+        CarbonTablePath.getSegmentFilesLocation(carbonTable.getTablePath());
+    List<String> segmentFiles = Arrays.stream(FileFactory.getCarbonFile(segmentFilesLocation)
+        .listFiles()).map(segmentFile -> segmentFile.getName()).collect(Collectors
+        .toList());
+    // there are no segments present in the Metadata folder. Can return here
+    if (segmentFiles.size() == 0) {
+      return;
+    }
+    ArrayList<String> staleSegments = new ArrayList<>(staleSegmentFiles.size());
```

Comment:
instead of creating new list, can you use stream.map and collect the list?

Reply:
done